coap/api/server.go

```diff
@@ -0,0 +1,126 @@
+package api
+
+import (
+	"fmt"
+	"net"
+	"strings"
+	"time"
+
+	mux "github.com/dereulenspiegel/coap-mux"
+	"github.com/mainflux/mainflux/coap"
+	manager "github.com/mainflux/mainflux/manager/client"
+
+	gocoap "github.com/dustin/go-coap"
+)
+
+func authKey(opt interface{}) (string, error) {
+	if opt == nil {
+		return "", errBadRequest
+	}
+	val, ok := opt.(string)
+	if !ok {
+		return "", errBadRequest
+	}
+	arr := strings.Split(val, "=")
+	if len(arr) != 2 || strings.ToLower(arr[0]) != "key" {
+		return "", errBadOption
+	}
+	return arr[1], nil
+}
+
+func authorize(msg *gocoap.Message, res *gocoap.Message, cid string) (publisher string, err error) {
+	// Device Key is passed as Uri-Query parameter, which option ID is 15 (0xf).
+	key, err := authKey(msg.Option(gocoap.URIQuery))
+	if err != nil {
+		switch err {
+		case errBadOption:
+			res.Code = gocoap.BadOption
+		case errBadRequest:
+			res.Code = gocoap.BadRequest
+		}
+		return
+	}
+
+	publisher, err = auth.CanAccess(cid, key)
+	if err != nil {
+		switch err {
+		case manager.ErrServiceUnreachable:
+			res.Code = gocoap.ServiceUnavailable
+		default:
+			res.Code = gocoap.Unauthorized
+		}
+	}
+	return
+}
+
+func serve(svc coap.Service, conn *net.UDPConn, data []byte, addr *net.UDPAddr, rh gocoap.Handler) {
+	msg, err := gocoap.ParseMessage(data)
+	if err != nil {
+		return
+	}
+	res := &gocoap.Message{
+		Type:      gocoap.NonConfirmable,
+		Code:      gocoap.Content,
+		MessageID: msg.MessageID,
+		Token:     msg.Token,
+		Payload:   []byte{},
+	}
+	switch msg.Type {
+	case gocoap.Reset:
+		cid := mux.Var(&msg, "id")
+		res.Type = gocoap.Acknowledgement
+		publisher, err := authorize(&msg, res, cid)
+		if err != nil {
+			res.Code = gocoap.Unauthorized
+		} else {
```

Comment:
From [Effective Go](https://golang.org/doc/effective_go.html#if): `In the Go libraries, you'll find that when an if statement doesn't flow into the next statement—that is, the body ends in break, continue, goto, or return—the unnecessary else is omitted.`

Reply:
I agree with this. Please omit `else`s wherever possible.